text/0779-first-class-component-templates.md

```diff
@@ -0,0 +1,1401 @@
+---
+Stage: Accepted
+Start Date: 2021-12-03
+Release Date: Unreleased
+Release Versions:
+  ember-source: vX.Y.Z
+  ember-data: vX.Y.Z
+Relevant Team(s): Ember.js, Learning, Ember CLI
+RFC PR: https://github.com/emberjs/rfcs/pull/779
+---
+
+
+# First Class Component Templates
+
+
+## Summary
+
+Adopt `<template>` tags as a format for making component templates first-class participants in JavaScript and TypeScript with [strict mode][rfc-0496] template semantics. In support of the new syntax, adopt new custom JavaScript and TypeScript files with the extensions `.gjs` and `.gts` respectively.
+
+First class component templates address a number of pain points in today’s component authoring world, and providing a number of new capabilities to Ember and Glimmer users:
+
+- accessing local JavaScript values with no ceremony and no backing class, enabling much easier use of existing JavaScript ecosystem tools, including especially styling libraries—standard [CSS Modules][css-modules] will “just work,” for example
+
+- authoring more than one component in a single file, where colocation makes sense—and thereby providing more control over a component’s public API
+
+- likewise authoring locally-scoped helpers, modifiers, and other JavaScript functionality
+
+First class component templates offer these new capabilities while not only maintaining but *improving* Ember’s long-standing commitment to integrated testing, in that it allows app and test code to share a single authoring paradigm—substantially simplifying our teaching story. Similarly, it preserves Ember’s long-standing commitment to treating JavaScript and HTML (and CSS!) as distinctive concerns which, however closely related, are not the *same*.
+
+<details><summary>Full-fledged example showing how this might work in practice</summary>
+
+(Note that for this and all following examples, I assume [RFC #0757: Default Modifier Manager][rfc-0757] for simplicity, but it does not meaningfully change *this* proposal.)
+
+[rfc-0757]: https://github.com/emberjs/rfcs/pull/757
+
+```js
+import Component from '@glimmer/component';
+import { tracked } from '@glimmer/tracking';
+
+const Greet = <template>
```

Comment:
This is tangential to the RFC, but the lack of symmetry in the starting and ending template tags have been bothering me.

I've started writing components like this:
```jsx
const Greet =
  <template>
    <p>Hello, {{@name}}!</p>
  </template>
```
Thoughts?

Reply:
I couldn’t possibly care less: I share the preference for symmetry but dislike the extra indentation and have many other such small tastes. Whatever we standardize on will be fine, and I will leave those arguments to bikeshedding in the Prettier repo. 😂

Reply:
Both options feel readable to me ✌🏻